In [1]:
import json
from typing import Annotated

import nest_asyncio
import requests
import uvicorn
from fastapi import FastAPI, Path, Query
from pydantic import BaseModel

<h1 style="text-align:center; color:#005bbd; font-size:30px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">Multiple Body</h1>

In [2]:
class Item(BaseModel):
    name: str
    description: str | None = None
    price: float
    tax: float | None = None


class User(BaseModel):
    username: str
    full_name: str | None = None

In [6]:
app = FastAPI()


@app.get("/index/")
async def route(item: Item, user: User):
    return {"item": item, "user": user}


if __name__ == "__main__":
    nest_asyncio.apply()
    uvicorn.run(app, port=4444)

In [8]:
item_data = {"name": "Mubeen", "description": "Hello World", "price": 15.8, "tax": 5.8}
user_data = {"username": "mubeen123", "full_name": "Mubeen Ahmad"}

In [29]:
req = requests.get(
    "http://127.0.0.1:4444/index", json={"item": item_data, "user": user_data}
)
response = req.content.decode("utf-8")
json.loads(response)

{'item': {'name': 'Mubeen',
  'description': 'Hello World',
  'price': 15.8,
  'tax': 5.8},
 'user': {'username': 'mubeen123', 'full_name': 'Mubeen Ahmad'}}

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body style="font-family: Sans-serif;">
    <div style="color: green; font-size: 18px; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black; border: 8px double black; padding: 20px;">
        <h2>Handling Multiple Pydantic Models in FastAPI</h2>
        <p>Yeh example demonstrate karta hai kaise multiple Pydantic models ko FastAPI mein body parameters ke roop mein handle kiya jata hai.</p>
        <ul> <h3><li><code>class Item(BaseModel)</code></li></h3>
            <h4 style="color: red;">
                <p>Item model define karta hai:</p>
                <ul>
                    <li><strong>name</strong>: Required string field</li>
                    <li><strong>description</strong>: Optional string field with default value <code>None</code></li>
                    <li><strong>price</strong>: Required float field</li>
                    <li><strong>tax</strong>: Optional float field with default value <code>None</code></li>
                </ul>
            </h4>
        </ul>
        <hr style="margin: 20px 0;">
        <ul>
            <h3><li><code>class User(BaseModel)</code></li></h3>
            <h4 style="color: red;">
                <p>User model define karta hai:</p>
                <ul>
                    <li><strong>username</strong>: Required string field</li>
                    <li><strong>full_name</strong>: Optional string field with default value <code>None</code></li></ul>
            </h4>
        </ul>
        <hr style="margin: 20px 0;">
        <ul>
            <h3><li><code>@app.get("/index/")</code></li></h3>
            <h4 style="color: red;">
                <p>Yeh endpoint <code>/index/</code> do parameters accept karta hai: <code>item</code> jo <code>Item</code> type ka hai aur <code>user</code> jo <code>User</code> type ka hai.</p>
            </h4>
        </ul>
        <hr style="margin: 20px 0;">
        <ul>
            <h3><li><code>Making a Request</code></li></h3>
            <h4 style="color: red;">
                <p>Example request data:</p>
                <pre>
item_data = {"name": 'Mubeen', "description": 'Hello World', "price": 15.8, "tax": 5.8}
user_data = {'username': 'mubeen123', 'full_name': 'Mubeen Ahmad'}
                </pre>
                <p>GET request to the endpoint:</p>
                <pre>
req = requests.get("http://127.0.0.1:4444/index", json={'item': item_data, 'user': user_data})
response = req.content.decode("utf-8")
json.loads(response)
                </pre>
            </h4>
        </ul>
        <hr style="margin: 20px 0;">
        <ul>
            <h3><li><code>Note:</code></li></h3>
            <h4 style="color: red;">Yahan, <code>item</code> aur <code>user</code> parameters ke names aur structure exactly match hone chahiye jo function <code>route</code> mein pass kiya gaya hai.</h4>
        </ul>
    </div>
</body>
</html>


<h1 style="text-align:center; color:#005bbd; font-size:30px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">Singular values in body (Body() parameter)</h1>


In [4]:
from fastapi import Body

In [12]:
app = FastAPI()


class User(BaseModel):
    username: str
    full_name: str | None = None


@app.post("/query/")
async def query_exp(user: User, importance: int = None):
    return (user, importance)


@app.post("/body/")
async def body_exp(user: User, importance: int = Body()):
    return (user, importance)


if __name__ == "__main__":
    nest_asyncio.apply()
    uvicorn.run(app, port=4444)

In [14]:
data = {"username": "mubeen", "full_name": "Mubeen Ahmad"}

req = requests.post("http://127.0.0.1:4444/query/?importance=123", json=data)
req.content.decode("utf-8")

'[{"username":"mubeen","full_name":"Mubeen Ahmad"},123]'

In [15]:
data = {
    "user": {"username": "mubeen", "full_name": "Mubeen Ahmad"},
    "importance": "123",
}
req = requests.post("http://127.0.0.1:4444/body/", json=data)
req.content.decode("utf-8")

'[{"username":"mubeen","full_name":"Mubeen Ahmad"},123]'

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body style="font-family: Sans-serif;">
    <div style="color: green; font-size: 18px; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black; border: 8px double black; padding: 20px;">
        <h2>Using Body Class in FastAPI</h2>
        <p>Yeh example FastAPI mein query parameters aur request body ke use ko demonstrate karta hai. Saath hi Body class ka istemal singular values ko request body mein pass karne ke liye dikhata hai.</p>
        <ul>
            <h3><li><code>/query/ Endpoint:</code></li></h3>
            <h4 style="color: red;">
                <p>Yeh endpoint query parameters aur request body ka istemal karta hai.</p>
                <p><strong>Example GET Request:</strong> <code>/query/?importance=123</code></p>
            </h4>
        </ul>
        <hr style="margin: 20px 0;">
        <ul>
            <h3><li><code>/body/ Endpoint:</code></li></h3>
            <h4 style="color: red;">
                <p>Yeh endpoint request body ke singular values ko handle karta hai, sath hi Body class ka istemal karta hai.</p>
                <p><strong>Example POST Request:</strong></p>
                <pre>
data = {'user': {'username': 'mubeen', 'full_name': 'Mubeen Ahmad'}, 'importance': '123'}
req = requests.post("http://127.0.0.1:4444/body/", json=data)
                </pre>
            </h4>
        </ul>
    </div>
</body>
</html>


<h1 style="text-align:center; color:#005bbd; font-size:30px; font-family:Sans-serif; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px Black;">Fields</h1>


In [18]:
from pydantic import BaseModel, Field

class Item(BaseModel):
    name: str
    price: float = Field(..., gt=0, description="The price must be greater than zero")
    description: str = Field(None, title="Item Description", max_length=50)

item1 = Item(name="Laptop", price=1200.50, description="High-performance laptop")
print(item1.model_dump())

{'name': 'Laptop', 'price': 1200.5, 'description': 'High-performance laptop'}


In [ ]:
app = FastAPI()


class Item(BaseModel):
    name: str
    price: float = Field(..., gt=0, description="The price must be greater than zero")
    description: str = Field(None, title="Item Description", max_length=50)

@app.post("/index/")
async def index(data : Item):
    return data


if __name__ == "__main__":
    nest_asyncio.apply()
    uvicorn.run(app, port=4444)

In [19]:
data = {"name" : "Laptop", "price" : 1200.50, "description" : "High-performance laptop"}
req = requests.post("http://127.0.0.1:4444/index/", json=data)
req.content.decode("utf-8")

'{"name":"Laptop","price":1200.5,"description":"High-performance laptop"}'

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body style="font-family: Arial, sans-serif; color: #333;">
     <div style="color: green; font-size: 18px; font-style: oblique; text-shadow: 0 0 3px white, 0 0 1px black; border: 8px double black; padding: 20px;">
        <h1 style="color: red; font-size:25px;">Field Pydantic Class</h1>
        <p style="margin-bottom: 16px;">Field Pydantic class Pydantic models ke individual fields ko customize karne ke liye use hota hai. Yeh fields ke liye validation rules aur metadata specify karne mein madad karta hai.</p>
        <h2  style="color: red; font-size:25px;">Features of Field Pydantic Class:</h2>
        <ul>
            <li><strong>Custom Validation:</strong> Validation rules jaise ki min/max values, regex patterns, etc. specify karne ke liye Field ka use hota hai.</li>
            <li><strong>Default Values:</strong> Fields ke liye default values specify karne ke liye Field ka use kiya ja sakta hai.</li>
            <li><strong>Metadata:</strong> Fields ke sath additional metadata jaise ki description, alias, title, etc. specify karne ke liye Field ka use kiya ja sakta hai, jo documentation ke liye madadgar hota hai.</li>
            <li><strong>Deprecation:</strong> Fields ko deprecated mark karne ke liye Field ka use kiya ja sakta hai.</li>
        </ul>
        <h2 style="color: red; font-size:25px;">Example Usage:</h2>
        <pre style="background-color: #f4f4f4; padding: 10px; border-radius: 5px; margin-bottom: 20px;">
            <code>
from pydantic import BaseModel, Field

class Item(BaseModel):
    name: str
    price: float = Field(..., gt=0, description="The price must be greater than zero")
    description: str = Field(None, title="Item Description", max_length=50)

item1 = Item(name="Laptop", price=1200.50, description="High-performance laptop")
print(item1.model_dump()) 
<b style="margin-bottom: 16px; color:red;" ><br>{'name': 'Laptop', 'price': 1200.5, 'description': 'High-performance laptop'}</b><br>
<code>
ab iska use Route me karty hn

app = FastAPI()

class Item(BaseModel):
    name: str
    price: float = Field(..., gt=0, description="The price must be greater than zero")
    description: str = Field(None, title="Item Description", max_length=50)

@app.post("/index/")
async def index(data : Item):
    return data
    
data = {"name" : "Laptop", "price" : 1200.50, "description" : "High-performance laptop"}
req = requests.post("http://127.0.0.1:4444/index/", json=data)
req.content.decode("utf-8")
            </code>
        </pre>
    </div>
</body>
</html>
